File: compat/suites/go-sdk/internal/groups/eventbridge.go
package groups

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/eventbridge"
	"github.com/aws/aws-sdk-go-v2/service/eventbridge/types"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	sqstypes "github.com/aws/aws-sdk-go-v2/service/sqs/types"
	"github.com/overcast-sh/overcast-compat-go-sdk/internal/clients"
	"github.com/overcast-sh/overcast-compat-go-sdk/internal/harness"
)

func EventBridge(c *clients.Clients) ServiceGroup {
	g := &ebGroup{c: c}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"eventbridge-target-fanout:PutFanoutTargets":              g.PutFanoutTargets,
			"eventbridge-target-fanout:PutEventsToQueueTarget":        g.PutEventsToQueueTarget,
			"eventbridge-target-fanout:PutEventsWithInputTransformer": g.PutEventsWithInputTransformer,
			"eventbridge-patterns:TestEventPattern":                   g.TestEventPattern,
			"eventbridge-patterns:TestEventPatternNoMatch":            g.TestEventPatternNoMatch,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"eventbridge-target-fanout": g.setupFanout,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"eventbridge-target-fanout": g.teardownFanout,
		},
	}
}

type ebGroup struct{ c *clients.Clients }

func (g *ebGroup) cl() *eventbridge.Client { return g.c.EventBridge() }

// ── eventbridge-patterns ──────────────────────────────────────────────────────

// patternsEvent is the envelope both pattern tests evaluate: every field AWS
// documents as mandatory for TestEventPattern is present.
func (g *ebGroup) patternsEvent(t *harness.TestContext) string {
	return fmt.Sprintf(`{"id":%q,"detail-type":"order.created","source":"compat.eventbridge-patterns",`+
		`"account":"000000000000","time":"2026-01-01T00:00:00Z","region":%q,"resources":[],"detail":{"orderId":"1"}}`,
		t.RunID, t.Region)
}

func (g *ebGroup) testEventPattern(ctx context.Context, t *harness.TestContext, pattern string) (bool, error) {
	resp, err := g.cl().TestEventPattern(ctx, &eventbridge.TestEventPatternInput{
		EventPattern: aws.String(pattern),
		Event:        aws.String(g.patternsEvent(t)),
	})
	if err != nil {
		return false, err
	}
	return resp.Result, nil
}

func (g *ebGroup) TestEventPattern(ctx context.Context, t *harness.TestContext) error {
	matched, err := g.testEventPattern(ctx, t, `{"source":["compat.eventbridge-patterns"],"detail-type":["order.created"]}`)
	if err != nil {
		return err
	}
	if !matched {
		return fmt.Errorf("TestEventPattern: expected Result=true for a matching pattern, got false")
	}
	return nil
}

func (g *ebGroup) TestEventPatternNoMatch(ctx context.Context, t *harness.TestContext) error {
	matched, err := g.testEventPattern(ctx, t, `{"source":["compat.eventbridge-patterns.other"]}`)
	if err != nil {
		return err
	}
	if matched {
		return fmt.Errorf("TestEventPatternNoMatch: expected Result=false for a non-matching pattern, got true")
	}
	return nil
}

// ── eventbridge-target-fanout ─────────────────────────────────────────────────
//
// Target fan-out: an event put on a bus reaches the rule's targets, with the
// target's input transformation applied first. The group provisions its own
// queues and rule so it can run independently of the other EventBridge groups.

func (g *ebGroup) setupFanout(ctx context.Context, t *harness.TestContext) error {
	sq := g.c.SQS()
	plain := fmt.Sprintf("oc-%s-fanout-plain", t.RunID)
	shaped := fmt.Sprintf("oc-%s-fanout-shaped", t.RunID)

	for key, name := range map[string]string{"eb_fanout_plain": plain, "eb_fanout_shaped": shaped} {
		created, err := sq.CreateQueue(ctx, &sqs.CreateQueueInput{QueueName: aws.String(name)})
		if err != nil {
			return err
		}
		attrs, err := sq.GetQueueAttributes(ctx, &sqs.GetQueueAttributesInput{
			QueueUrl:       created.QueueUrl,
			AttributeNames: []sqstypes.QueueAttributeName{"QueueArn"},
		})
		if err != nil {
			return err
		}
		t.Set(key+"_url", aws.ToString(created.QueueUrl))
		t.Set(key+"_arn", attrs.Attributes["QueueArn"])
	}

	source := fmt.Sprintf("oc.fanout.%s", t.RunID)
	rule := fmt.Sprintf("oc-%s-fanout", t.RunID)
	if _, err := g.cl().PutRule(ctx, &eventbridge.PutRuleInput{
		Name:         aws.String(rule),
		EventPattern: aws.String(fmt.Sprintf(`{"source":["%s"]}`, source)),
		State:        types.RuleStateEnabled,
	}); err != nil {
		return err
	}
	t.Set("eb_fanout_rule", rule)
	t.Set("eb_fanout_source", source)
	return nil
}

func (g *ebGroup) teardownFanout(ctx context.Context, t *harness.TestContext) error {
	if rule := t.GetString("eb_fanout_rule"); rule != "" {
		g.cl().RemoveTargets(ctx, &eventbridge.RemoveTargetsInput{ //nolint:errcheck
			Rule: aws.String(rule), Ids: []string{"plain", "shaped"},
		})
		g.cl().DeleteRule(ctx, &eventbridge.DeleteRuleInput{Name: aws.String(rule)}) //nolint:errcheck
	}
	for _, key := range []string{"eb_fanout_plain_url", "eb_fanout_shaped_url"} {
		if url := t.GetString(key); url != "" {
			g.c.SQS().DeleteQueue(ctx, &sqs.DeleteQueueInput{QueueUrl: aws.String(url)}) //nolint:errcheck
		}
	}
	return nil
}

func (g *ebGroup) PutFanoutTargets(ctx context.Context, t *harness.TestContext) error {
	rule := t.GetString("eb_fanout_rule")
	resp, err := g.cl().PutTargets(ctx, &eventbridge.PutTargetsInput{
		Rule: aws.String(rule),
		Targets: []types.Target{
			{Id: aws.String("plain"), Arn: aws.String(t.GetString("eb_fanout_plain_arn"))},
			{
				Id:  aws.String("shaped"),
				Arn: aws.String(t.GetString("eb_fanout_shaped_arn")),
				InputTransformer: &types.InputTransformer{
					InputPathsMap: map[string]string{"order": "$.detail.orderId"},
					InputTemplate: aws.String(`{"order":"<order>"}`),
				},
			},
		},
	})
	if err != nil {
		return err
	}
	if resp.FailedEntryCount > 0 {
		return fmt.Errorf("PutFanoutTargets: %d failed entries", resp.FailedEntryCount)
	}

	listed, err := g.cl().ListTargetsByRule(ctx, &eventbridge.ListTargetsByRuleInput{Rule: aws.String(rule)})
	if err != nil {
		return err
	}
	if len(listed.Targets) != 2 {
		return fmt.Errorf("PutFanoutTargets: rule has %d targets, want 2", len(listed.Targets))
	}
	for _, target := range listed.Targets {
		if aws.ToString(target.Id) != "shaped" {
			continue
		}
		if target.InputTransformer == nil || aws.ToString(target.InputTransformer.InputTemplate) == "" {
			return fmt.Errorf("PutFanoutTargets: InputTransformer did not round-trip")
		}
	}
	return nil
}

func (g *ebGroup) PutEventsToQueueTarget(ctx context.Context, t *harness.TestContext) error {
	if err := g.putFanoutEvent(ctx, t, "queue-target"); err != nil {
		return err
	}
	body, err := g.awaitFanoutMessage(ctx, t.GetString("eb_fanout_plain_url"), "queue-target")
	if err != nil {
		return fmt.Errorf("PutEventsToQueueTarget: %w", err)
	}
	if !strings.Contains(body, t.GetString("eb_fanout_source")) || !strings.Contains(body, "queue-target") {
		return fmt.Errorf("PutEventsToQueueTarget: delivered body missing the event envelope: %s", body)
	}
	return nil
}

func (g *ebGroup) PutEventsWithInputTransformer(ctx context.Context, t *harness.TestContext) error {
	if err := g.putFanoutEvent(ctx, t, "transformed"); err != nil {
		return err
	}
	body, err := g.awaitFanoutMessage(ctx, t.GetString("eb_fanout_shaped_url"), "transformed")
	if err != nil {
		return fmt.Errorf("PutEventsWithInputTransformer: %w", err)
	}
	if body != `{"order":"transformed"}` {
		return fmt.Errorf("PutEventsWithInputTransformer: delivered body = %s, want the rendered template", body)
	}
	return nil
}

func (g *ebGroup) putFanoutEvent(ctx context.Context, t *harness.TestContext, orderID string) error {
	resp, err := g.cl().PutEvents(ctx, &eventbridge.PutEventsInput{
		Entries: []types.PutEventsRequestEntry{{
			Source:     aws.String(t.GetString("eb_fanout_source")),
			DetailType: aws.String("FanoutTest"),
			Detail:     aws.String(fmt.Sprintf(`{"orderId":%q}`, orderID)),
		}},
	})
	if err != nil {
		return err
	}
	if resp.FailedEntryCount > 0 {
		return fmt.Errorf("PutEvents: %d failed entries", resp.FailedEntryCount)
	}
	return nil
}

// awaitFanoutMessage polls a queue for a delivered message carrying want.
// Both targets hang off one rule, so every event reaches both queues and a
// queue may hold an earlier test's message; matching on want (and consuming
// what does not match) keeps the tests order-independent. Delivery is
// asynchronous, so a bounded poll replaces a fixed sleep.
func (g *ebGroup) awaitFanoutMessage(ctx context.Context, queueURL, want string) (string, error) {
	for attempt := 0; attempt < 15; attempt++ {
		resp, err := g.c.SQS().ReceiveMessage(ctx, &sqs.ReceiveMessageInput{
			QueueUrl:            aws.String(queueURL),
			MaxNumberOfMessages: 10,
			WaitTimeSeconds:     1,
		})
		if err != nil {
			return "", err
		}
		var matched string
		for _, message := range resp.Messages {
			g.c.SQS().DeleteMessage(ctx, &sqs.DeleteMessageInput{ //nolint:errcheck
				QueueUrl: aws.String(queueURL), ReceiptHandle: message.ReceiptHandle,
			})
			if body := aws.ToString(message.Body); strings.Contains(body, want) {
				matched = body
			}
		}
		if matched != "" {
			return matched, nil
		}
		time.Sleep(100 * time.Millisecond)
	}
	return "", fmt.Errorf("no message containing %q delivered to the target queue", want)
}

File: compat/suites/python-sdk/groups/eventbridge.py
"""
groups/eventbridge.py — EventBridge compatibility test implementations.

Note: boto3 uses the service name "events" for EventBridge.
"""

from __future__ import annotations
import json
import time
from lib.harness import TestContext
from lib.clients import make_clients


def _eb(ctx: TestContext):
    # boto3 uses "events" as the service identifier for EventBridge
    return make_clients(ctx.endpoint, ctx.region).eventbridge


# ── eventbridge-target-fanout ─────────────────────────────────────────────────
#
# Target fan-out: an event put on a bus reaches the rule's targets, with the
# target's input transformation applied first. The group provisions its own
# queues and rule so it never races the other EventBridge groups (issue #388).


def _fanout_source(ctx: TestContext) -> str:
    return f"oc.fanout.{ctx.run_id}"


def _fanout_rule(ctx: TestContext) -> str:
    return f"oc-{ctx.run_id}-fanout"


def setup_eventbridge_target_fanout(ctx: TestContext) -> None:
    eb = _eb(ctx)
    sqs = make_clients(ctx.endpoint, ctx.region).sqs
    for key, name in (
        ("eb_fanout_plain", f"oc-{ctx.run_id}-fanout-plain"),
        ("eb_fanout_shaped", f"oc-{ctx.run_id}-fanout-shaped"),
    ):
        url = sqs.create_queue(QueueName=name)["QueueUrl"]
        attrs = sqs.get_queue_attributes(QueueUrl=url, AttributeNames=["QueueArn"])
        ctx[key + "_url"] = url
        ctx[key + "_arn"] = attrs["Attributes"]["QueueArn"]

    eb.put_rule(
        Name=_fanout_rule(ctx),
        EventPattern=json.dumps({"source": [_fanout_source(ctx)]}),
        State="ENABLED",
    )


def teardown_eventbridge_target_fanout(ctx: TestContext) -> None:
    eb = _eb(ctx)
    sqs = make_clients(ctx.endpoint, ctx.region).sqs
    try:
        eb.remove_targets(Rule=_fanout_rule(ctx), Ids=["plain", "shaped"])
    except Exception:
        pass
    try:
        eb.delete_rule(Name=_fanout_rule(ctx))
    except Exception:
        pass
    for key in ("eb_fanout_plain_url", "eb_fanout_shaped_url"):
        url = ctx.get(key)
        if not url:
            continue
        try:
            sqs.delete_queue(QueueUrl=url)
        except Exception:
            pass


def PutFanoutTargets(ctx: TestContext) -> None:
    eb = _eb(ctx)
    rule = _fanout_rule(ctx)
    resp = eb.put_targets(
        Rule=rule,
        Targets=[
            {"Id": "plain", "Arn": ctx["eb_fanout_plain_arn"]},
            {
                "Id": "shaped",
                "Arn": ctx["eb_fanout_shaped_arn"],
                "InputTransformer": {
                    "InputPathsMap": {"order": "$.detail.orderId"},
                    "InputTemplate": '{"order":"<order>"}',
                },
            },
        ],
    )
    if resp.get("FailedEntryCount", 0) > 0:
        raise AssertionError(f"PutFanoutTargets: {resp['FailedEntryCount']} failed entries {resp.get('FailedEntries')}")

    targets = eb.list_targets_by_rule(Rule=rule).get("Targets", [])
    if len(targets) != 2:
        raise AssertionError(f"PutFanoutTargets: rule has {len(targets)} targets, want 2")
    shaped = next((t for t in targets if t["Id"] == "shaped"), None)
    if not shaped or not shaped.get("InputTransformer", {}).get("InputTemplate"):
        raise AssertionError("PutFanoutTargets: InputTransformer did not round-trip")


def PutEventsToQueueTarget(ctx: TestContext) -> None:
    _put_fanout_event(ctx, "queue-target")
    body = _await_fanout_message(ctx, ctx["eb_fanout_plain_url"], "queue-target")
    if _fanout_source(ctx) not in body or "queue-target" not in body:
        raise AssertionError(f"PutEventsToQueueTarget: delivered body missing the event envelope: {body}")


def PutEventsWithInputTransformer(ctx: TestContext) -> None:
    _put_fanout_event(ctx, "transformed")
    body = _await_fanout_message(ctx, ctx["eb_fanout_shaped_url"], "transformed")
    if body != '{"order":"transformed"}':
        raise AssertionError(f"PutEventsWithInputTransformer: delivered body = {body}, want the rendered template")


def _put_fanout_event(ctx: TestContext, order_id: str) -> None:
    resp = _eb(ctx).put_events(
        Entries=[
            {
                "Source": _fanout_source(ctx),
                "DetailType": "FanoutTest",
                "Detail": json.dumps({"orderId": order_id}),
            }
        ]
    )
    if resp.get("FailedEntryCount", 0) > 0:
        raise AssertionError(f"PutEvents: {resp['FailedEntryCount']} failed entries")


def _await_fanout_message(ctx: TestContext, queue_url: str, want: str) -> str:
    """Poll a target queue for a delivered message containing ``want``.

    Both targets hang off one rule, so every event reaches both queues and a
    queue may hold an earlier test's message; matching on ``want`` (and
    consuming what does not match) keeps the tests order-independent.
    Delivery is asynchronous, so a bounded poll replaces a fixed sleep.
    """
    sqs = make_clients(ctx.endpoint, ctx.region).sqs
    for _ in range(15):
        resp = sqs.receive_message(QueueUrl=queue_url, MaxNumberOfMessages=10, WaitTimeSeconds=1)
        matched = None
        for message in resp.get("Messages", []):
            try:
                sqs.delete_message(QueueUrl=queue_url, ReceiptHandle=message["ReceiptHandle"])
            except Exception:
                pass
            if want in message["Body"]:
                matched = message["Body"]
        if matched is not None:
            return matched
        time.sleep(0.1)
    raise AssertionError(f"no message containing {want!r} delivered to the target queue")


# ── eventbridge-patterns ──────────────────────────────────────────────────────
#
# Stateless: TestEventPattern evaluates a pattern against an event without any
# bus, rule, or target involved, so this group needs no setup/teardown.


def _pattern_test_event(ctx: TestContext) -> str:
    return json.dumps({
        "id": ctx.run_id,
        "detail-type": "order.created",
        "source": "compat.eventbridge-patterns",
        "account": "000000000000",
        "time": "2026-01-01T00:00:00Z",
        "region": ctx.region,
        "resources": [],
        "detail": {"orderId": "1"},
    })


def TestEventPattern(ctx: TestContext) -> None:
    eb = _eb(ctx)
    pattern = json.dumps({
        "source": ["compat.eventbridge-patterns"],
        "detail-type": ["order.created"],
    })
    resp = eb.test_event_pattern(EventPattern=pattern, Event=_pattern_test_event(ctx))
    if resp.get("Result") is not True:
        raise AssertionError(f"TestEventPattern: expected Result=True, got {resp.get('Result')!r}")


def TestEventPatternNoMatch(ctx: TestContext) -> None:
    eb = _eb(ctx)
    pattern = json.dumps({"source": ["compat.eventbridge-patterns.other"]})
    resp = eb.test_event_pattern(EventPattern=pattern, Event=_pattern_test_event(ctx))
    if "Result" not in resp:
        raise AssertionError(f"TestEventPatternNoMatch: missing Result in {resp}")
    if resp["Result"] is not False:
        raise AssertionError(f"TestEventPatternNoMatch: expected Result=False, got {resp['Result']!r}")


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "eventbridge-target-fanout:PutFanoutTargets": PutFanoutTargets,
    "eventbridge-target-fanout:PutEventsToQueueTarget": PutEventsToQueueTarget,
    "eventbridge-target-fanout:PutEventsWithInputTransformer": PutEventsWithInputTransformer,
    "eventbridge-patterns:TestEventPattern": TestEventPattern,
    "eventbridge-patterns:TestEventPatternNoMatch": TestEventPatternNoMatch,
}

SETUP = {
    "eventbridge-target-fanout": setup_eventbridge_target_fanout,
}

TEARDOWN = {
    "eventbridge-target-fanout": teardown_eventbridge_target_fanout,
}
